automated integration tests for cloudstack

resmo opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
tests

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.3
```

##### SUMMARY
automated integration tests for cloudstack

/cc @gundalow @mattclay

gundalow commented:
Hi,
Thanks for this. Not managed to catch you online at the same point as mattclay (West Coast USA)

How slow is it to bring this up in Docker? If it's only a few minutes then we can add that into Shippable, rather than requiring external resources (something we are trying to remove).

The tests will need moving from `integration/roles` to `integration/targets`

gundalow commented:
@resmo When you get a chance could you have a look at my comments above.

resmo commented:
@gundalow yes, moving them to targets should not be a big deal. But I am currently too busy. I should get more time next week though.

resmo commented:
@gundalow @mattclay  I have done some work and have a working prove of work. 

Still room left to improve and I would need some help to integrate it properly into the new best practice of integration testing :)

Look at the cloudstack.sh to get started. It bascically starts a docker image and a playbook, having some roles. one role is the test_cs_prepare, in which I do all the authentication preparation. The rest is self explainable. 

On my machine it takes about 6 minutes to start the setup. I currently only run one integration test `test_cs_user` (see https://github.com/ansible/ansible/pull/20552/commits/af707bdb0b911406d6a3e387666141ca49d8f173#diff-fd35e8e13de0ec5189e30a883b8f2b3cR10)

Feedback appreciated :)

mattclay commented:
@resmo I'd like to see the CloudStack tests as first class test targets, just like the rest of our integration tests. I'm working on a few ideas on how to do that cleanly.

mattclay commented:
I'd like to see the Dockerfile added to our repo, the same as the others we use for testing. We can wait to do that until we get closer to having everything working.

resmo commented:
we are now at 9min 10s, only for the tests (631 tasks) to run.

mattclay commented:
@resmo Could you open a PR to add the `cloudstack-sim` `Dockerfile` and dependencies to `test/utils/docker/cloudstack-sim/`? I'd like to be able to build it from the Ansible Docker Hub account so I can reference the image as `ansible/ansible:cloudstack-sim` from within `ansible-test`.

mattclay commented:
@resmo I've deployed the initial support for testing CloudStack modules via ansible-test. I moved over some of the integration test roles to ansible-test targets. The remaining ones may require your changes from this PR before they can be migrated.

Here's the PR with the initial migration: https://github.com/ansible/ansible/pull/24317

resmo committed:
cloudstack: revamp integration tests for automated CI

resmo committed:
cloudstack: adjust dependency role

resmo committed:
cloudstack: test: revert unintentional py3 skip

resmo committed:
cloudstack: fix cs_pod, add missing zone param in setup absent

resmo committed:
cloudsatck: cs_prepare is not used anymore

resmo commented:
Yey, it passes :)

mattclay commented:
Test roles which were copied to the targets directory should be removed from the roles directory.

resmo commented:
Yes of course, cleaning up

resmo committed:
Revert "cloudsatck: cs_prepare is not used anymore"

This reverts commit 1f70826358b08f0d3324115e30a7dd1c13c0c695.

resmo committed:
cloudstack: tests: hide cs_prepare

resmo committed:
cloudstack: test: cleanup unused roles

resmo committed:
cloudstack: test: add cloudstack-simulator docker image setup

resmo committed:
cloudstack: test: adjust the role path to targets in Makefile

resmo commented:
* Cleaned up old role
* Kept the cloustack.yml in integration and changed the role path so the test can be easily run locally.
* Added the docker src for cloudstack-simulator

Anything else? ;)

mattclay commented:
You can run the tests locally with: `ansible-test integration cloud/cs/ -v`

If you don't want `ansible-test` to start the simulator for you, use the `test/integration/cloud-config-cs.ini.template` to tell it where to find the test endpoint.

Does that eliminate the need to maintain the legacy playbook and `make` command for you?

resmo commented:
Nice! In this case I'll clean up the Makefile. I see no need to keep it.

resmo committed:
cloudstack: test: clean up legacy test execution

mattclay commented:
If you do encounter issues running the tests via `ansible-test` locally, please let me know, as it's intended to be fully usable (and convenient) for both local testing and running the tests on Shippable.

mattclay commented on test/integration/targets/cs_instance/tasks/main.yml:
```diff
@@ -9,3 +9,12 @@
+#- include: absent.yml
+#  tags: test_cs_instance_adv
+#- include: cleanup.yml
+#  tags: test_cs_instance_adv
```
Why are these includes commented out?

resmo replied:
good question, I can't remember why. Let me verify if it works.

resmo replied:
I checked again. These tests are related to features I am developing regarding advanced networking. So these tests do not reflect the current state. I'll remove them for now and add them once this is merged along with the features.

mattclay commented on test/integration/targets/cs_prepare/defaults/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
```
There shouldn't be any need for `cs_prepare` with `ansible-test`. It takes care of the requirements when running with any of the options: `--tox`, `--docker` or `--requirements`, as well as configuring the environment and fetching the API keys from the simulator.

resmo replied:
ack, was used in legacy. removing.

mattclay commented on test/utils/docker/cloudstack-simulator/README.md:
```diff
@@ -0,0 +1,20 @@
+# CloudStack Simulator as Docker Image
```
For new documentation we prefer reStructuredText instead of Markdown.

resmo committed:
readme: md to rst

resmo committed:
cloudstack: tests: removing out commented tests for unimplemented features

resmo committed:
remove cs_prepare again, unused
